### DISCUSSION TITLE
Add a method of getting to osu!direct from song select

### DISCUSSION THREAD

**Discussion by Craftplacer** [ppy/osu#13107] [Ideas]
**Describe the new feature:** 
osu!lazer would show suggestions of downloadable beatmaps inside the beatmap carousel, they would either be half the opacity or have a little cloud icon, clicking on them or pressing enter, would show a progressbar under them and automatically start playing when downloaded. Missing difficulties will be also be shown, unless hidden by the player.

**Limitations:**
osu! shouldn't show them if the search entered is too broad (over 10-25 results, most likely searching an artist).
The alternative to this would be that osu!lazer only should show the first 5 most popular maps, with an entry linking to osu!direct.

**Proposal designs of the feature:** 
Normal:
![image](https://user-images.githubusercontent.com/22963120/73592707-3d0e8d80-44fd-11ea-9a58-a026a5386215.png)

Downloading:
![image](https://user-images.githubusercontent.com/22963120/73592698-27996380-44fd-11ea-8260-121bcc1cd97d.png)

**Comment by bdach** [comment]
I'm quite opposed to this, it feels really intrusive and one step removed from advertising to me. No need to stuff "automated recommendations" anywhere they could go.

**Comment by Craftplacer** [comment]
In my opinion, it can be optional of course, if the player searches for something they know of (e.g. a popular song), it would be neat if osu! could show existing beatmaps so the player doesn't have to open direct first.

But I do agree that osu! shouldn't recommend maps everywhere, the scope is intended just to be the search inside the song select.

**Comment by frenzibyte** [comment]
Seems unnecessary IMO because the osu!direct can be opened just by <kbd>Ctrl</kbd>+<kbd>D</kbd>, I don't see a difficulty with just searching there.

**Comment by Stedoss** [comment]
I think an alternative to this would be to possibly have an option to load the osu!direct overlay with the currently inputted local search text. So, when no local entries can be found you can bring up direct with what you were searching for locally. (Hopefully that made sense, and it's not already implemented lol)

**Comment by Craftplacer** [comment]
@mouzedrift's suggestion was to add something like this
![image](https://user-images.githubusercontent.com/22963120/73593314-0471b280-4503-11ea-889d-b4cecb0ccce1.png)
only if no beatmaps were found

**Comment by mcendu** [comment]
Maybe add an entry at the end of all beatmaps which says "Search online for beatmaps", which when clicked brings up the direct overlay searching for the specified name.

**EDIT**: ninja'd, I am in china so sometime the images won't load

**Comment by LastExceed** [comment]
I think we should completely rethink the way beatmap discovery and storage works and get rid of having 2 different places to search maps (direct and songselect) in the first place:

- Put all newly downloaded maps into a cache until the user explicitly decides to either save or delete them. Making this decision should be very accessible (a single click or keystroke if possible) and adding a beatmap to a collection should automatically count as saving
- Add a user option "delete all cached maps" to clean the database from all maps that haven't been saved explicitly
- Add a user option to automatically delete beatmaps after residing in the cache for some amount of time (e.g. a week), not sure if this should be on or off by default
- Keep a history of beatmap downloads, so in case the user forgot to save a beatmap and can't remember it's name, they can go back in the history to find and re-download it
- osu!direct is removed, instead the unordered view of song selection ("all" in lazer, "no grouping" in stable) has an option to include online results, eg a checkbox that is grayed out or hidden when you're offline or not a supporter
- if necessary, introduce a new search parameter "DownloadState=saved/cached/cloud" or similar

**Comment by peppy** [comment]
We also do have plans on adding a method of getting from song select to online results via a panel or other display, but this needs design / UX considerations. It would also display in cases the user has no beatmaps loaded, for instance.

I do have an eventual plan to add "cloud storage" of user's beatmap lists (maybe supporter only), which would download on-demand when a beatmap is accessed.
